diff --git a/functions b/functions
index b889c1a..6cc1943 100644
--- usr/share/flash-kernel/functions
+++ usr/share/flash-kernel/functions
@@ -710,6 +710,10 @@
 }
 
 main() {
+if systemd-detect-virt --chroot; then
+	echo "A chroot environment has been detected. Interrupt flash-kernel." >&2
+	exit 0
+fi
 force="no"
 if [ "x$1" = "x--force" ]; then
 	force="yes"
diff --git a/initramfs-tools/hooks/flash_kernel_set_root b/initramfs-tools/hooks/flash_kernel_set_root
index ac57d12..98d76ab 100755
--- usr/share/initramfs-tools/hooks/flash_kernel_set_root
+++ usr/share/initramfs-tools/hooks/flash_kernel_set_root
@@ -62,6 +62,11 @@ if systemd-detect-virt --quiet --container; then
 	exit 0
 fi
 
+# Do not run inside a chroot environment as well
+if systemd-detect-virt --chroot; then
+	exit 0
+fi
+
 # Record the root filesystem device for use during boot
 rootdev=$(egrep '^[^# 	]+[ 	]+/[ 	]' /etc/fstab | awk '{print $1}') || true
 
